OpenRA.Mods.Common/Activities/FindResources.cs

```diff
@@ -34,103 +37,112 @@ public FindResources(CPos avoidCell)
 
-			// Determine where to search from and how far to search:
-			var searchFromLoc = harv.LastOrderLocation ?? (harv.LastLinkedProc ?? harv.LinkedProc ?? self).Location;
-			var searchRadius = harv.LastOrderLocation.HasValue ? harvInfo.SearchFromOrderRadius : harvInfo.SearchFromProcRadius;
-			var searchRadiusSquared = searchRadius * searchRadius;
+			var closestHarvestablePosition = ClosestHarvestablePos(self, harv);
 
-			// Find harvestable resources nearby:
-			var path = self.World.WorldActor.Trait<IPathFinder>().FindPath(
-				PathSearch.Search(self.World, mobileInfo, self, true)
-					.WithHeuristic(loc =>
-					{
-						// Avoid this cell:
-						if (avoidCell.HasValue && loc == avoidCell.Value)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						// Don't harvest out of range:
-						var distSquared = (loc - searchFromLoc).LengthSquared;
-						if (distSquared > searchRadiusSquared)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost * 2;
-
-						// Get the resource at this location:
-						var resType = resLayer.GetResource(loc);
-						if (resType == null)
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						// Can the harvester collect this kind of resource?
-						if (!harvInfo.Resources.Contains(resType.Info.Name))
-							return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-
-						if (territory != null)
-						{
-							// Another harvester has claimed this resource:
-							ResourceClaim claim;
-							if (territory.IsClaimedByAnyoneElse(self, loc, out claim))
-								return EstimateDistance(loc, searchFromLoc) + Constants.CellCost;
-						}
-
-						return 0;
-					})
-					.FromPoint(self.Location));
-
-			if (path.Count == 0)
+			// If no harvestable position could be found, then return an empty path
+			if (closestHarvestablePosition == CPos.Zero)
 			{
 				if (!harv.IsEmpty)
 					return new DeliverResources();
+
+				// Get out of the way if we are
+				harv.UnblockRefinery(self);
+				var moveTo = harv.LastHarvestedCell ?? (self.Location + new CVec(0, 4));
+				self.QueueActivity(mobile.MoveTo(moveTo, 1));
+				self.SetTargetLine(Target.FromCell(self.World, moveTo), Color.Gray, false);
+
+				var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
+				if (NextActivity != null)
+					return Util.SequenceActivities(NextActivity, new Wait(randFrames), new FindResources());
 				else
-				{
-					// Get out of the way if we are:
-					harv.UnblockRefinery(self);
-					var randFrames = 125 + self.World.SharedRandom.Next(-35, 35);
-					if (NextActivity != null)
-						return Util.SequenceActivities(NextActivity, new Wait(randFrames), new FindResources());
-					else
-						return Util.SequenceActivities(new Wait(randFrames), new FindResources());
-				}
+					return Util.SequenceActivities(new Wait(randFrames), new FindResources());
 			}
 
 			// Attempt to claim a resource as ours:
 			if (territory != null)
```

Comment:
By working do you mean if it ever enter that if? By logic and reasoning it should not ever enter, but I didn't understand why that null check is there. But as it didn't harm I haven't tried to move it.

But once again, by logic it doesn't make sense.